resources/type-system/upper-lower-bounds.md

```diff
@@ -0,0 +1,390 @@
+         `T0 Function<X0 extends B10, ... Xm extends B1m>(P10, ... P1k, Named1)` = 
+   `R0 Function<X0 extends B20, ..., Xm extends B2m>(P20, ..., P2k, Named2)` if:
+     - each `B0i` and `B1i` are equal types (syntactically)
+     - All positional parameters are required
+     - `R0` is **UP**(`T0`, `T1`)
+     - `B2i` is `B0i`
+     - `P2i` is **DOWN**(`P0i`, `P1i`)
+     - `Named0` contains `R0i xi` if `R1i xi` is a required named parameter in `Named1`
+     - `Named1` contains `R1i xi` if `R0i xi` is a required named parameter in `Named0`
+     - `Named2` contains exactly `R2i xi` for each `xi` in both `Named0` and `Named1` 
+        - where `R0i xi` is in `Named0`
+        - where `R1i xi` is in `Named1`
+        - and `R2i` is **DOWN**(`R0i`, `R1i`)
+        - and `R2i xi` is required if `xi` is required in either `Named0` or `Named1`
+
+- **UP**(`T Function<...>(...)`, `S Function<...>(...)`) = `Function` otherwise
+- **UP**(`T1`, `T2`) = `T2` if `T1` <: `T2`
+  - Note that both types must be interface types at this point
+- **UP**(`T1`, `T2`) = `T1` if `T2` <: `T1`
+  - Note that both types must be interface types at this point
+- **UP**(`C<T0, ..., Tn>`, `C<S0, ..., Sn>`) = `C<R0,..., Rn>` where `Ri` is **UP**(`Ti`, `Si`)
+- **UP**(`C0<T0, ..., Tn>`, `C1<S0, ..., Sk>`) = least upper bound of two interfaces
+  as in Dart 1.
+
+## Lower bounds
+
+We define the lower bound of two types T1 and T2 to be **DOWN**(T1,T2) as
+follows.
+
+- **DOWN**(`T`, `T`) = `T`
+
+- **DOWN**(`T1`, `T2`) where **TOP**(`T1`) and **TOP**(`T2`) = 
+  - `T1` if **MORETOP**(`T2`, `T1`)
+  - `T2` otherwise
+- **DOWN**(`T1`, `T2`) = `T2` if **TOP**(`T1`)
+- **DOWN**(`T1`, `T2`) = `T1` if **TOP**(`T2`)
+
+- **DOWN**(`T1`, `T2`) where **BOTTOM**(`T1`) and **BOTTOM**(`T2`) =
+  - `T1` if **MOREBOTTOM**(`T1`, `T2`)
+  - `T2` otherwise
+- **DOWN**(`T1`, `T2`) = `T2` if **BOTTOM**(`T2`)
+- **DOWN**(`T1`, `T2`) = `T1` if **BOTTOM**(`T1`)
+
+
+- **DOWN**(`T1`, `T2`) where **NULL**(`T1`) and **NULL**(`T2`) =
+  - `T1` **MOREBOTTOM**(`T1`, `T2`)
+  - `T2` otherwise
+
+- **DOWN**(`T1`, `T2`) where **NULL**(`T1`) =
+  - `T1` if  `T2` is nullable
+  - `Never` otherwise
+
+- **DOWN**(`T1`, `T2`) where **NULL**(`T2`) =
+  - `T2` if  `T1` is nullable
+  - `Never` otherwise
+
+- **DOWN**(`T1`, `T2`) where **OBJECT**(`T1`) and **OBJECT**(`T2`) =
+  - `T1` if **MORETOP**(`T2`, `T1`)
+  - `T2` otherwise
+
+- **DOWN**(`T1`, `T2`) where **OBJECT**(`T1`) =
+  - `T2` if `T2` is non-nullable
+  - `Never` otherwise
+
+- **DOWN**(`T1`, `T2`) where **OBJECT**(`T2`) =
+  - `T2` if `T1` is non-nullable
+  - `Never` otherwise
+
+- **DOWN**(`T1*`, `T2*`) = `S*` where `S` is **DOWN**(`T1`, `T2`)
+- **DOWN**(`T1*`, `T2?`) = `S*` where `S` is **DOWN**(`T1`, `T2`)
+- **DOWN**(`T1?`, `T2*`) = `S*` where `S` is **DOWN**(`T1`, `T2`)
+- **DOWN**(`T1*`, `T2`) = `S` where `S` is **DOWN**(`T1`, `T2`)
+- **DOWN**(`T1`, `T2*`) = `S` where `S` is **DOWN**(`T1`, `T2`)
```

Comment:
Line 246, 247 forget the `*`. We could have `DOWN(int*, int?)` and get `int`, but it seems likely that we would want `int*`. WDYT?

Reply:
Well, the *, ? case is already handled in the lines above.  I don't know that there's a right answer here: if `S <: T2, S <: T1*` then `S* <: T2, S* <: T1*`, so either choice works.

Reply:
Thinking about this again, and in terms of existential types, I think it's fine as it is.